{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "69dedc7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30acffeb",
   "metadata": {},
   "source": [
    "# Mice methylation&RNA data - delete duplicate genes and save as .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "334ff9ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing data_methylation_hm450\n",
      "Original: (25883, 73)\n",
      "duplicated rows: 36\n",
      "Deduplicated: (25865, 73)\n",
      "Processing data_mrna_seq_rpkm\n",
      "Original: (19552, 66)\n",
      "duplicated rows: 8\n",
      "Deduplicated: (19548, 66)\n"
     ]
    }
   ],
   "source": [
    "for data_name in [\"data_methylation_hm450\", \"data_mrna_seq_rpkm\"]:\n",
    "    print(f\"Processing {data_name}\")\n",
    "    expression_data = pd.read_csv(f\"../data/Mice/{data_name}.txt\", index_col=0, sep='\\t')\n",
    "    if data_name == \"data_methylation_hm450\":\n",
    "        expression_data.drop(['Entrez_Gene_Id'], inplace=True, axis=1)\n",
    "\n",
    "\n",
    "    print(f\"Original: {expression_data.shape}\")\n",
    "\n",
    "    # Check duplicated rows\n",
    "    df = expression_data\n",
    "    duplicated_rows = df[df.index.duplicated(keep=False)]\n",
    "    print(f\"duplicated rows: {len(duplicated_rows)}\")\n",
    "\n",
    "    # Compute row means\n",
    "    row_means = df.mean(axis=1)\n",
    "\n",
    "    # Get best row per gene (index) by mean\n",
    "    best_rows = row_means.groupby(df.index).idxmax()\n",
    "\n",
    "    # Select those rows\n",
    "    df_deduped = df.loc[best_rows]\n",
    "\n",
    "    # Now force unique index — remove any accidental duplicate labels\n",
    "    df_deduped = df_deduped[~df_deduped.index.duplicated(keep='first')]\n",
    "\n",
    "    print(\"Deduplicated:\", df_deduped.shape)\n",
    "\n",
    "    # Save as CSV\n",
    "    df_deduped.to_csv(f'../data/Mice/{data_name}.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7d1fc5c",
   "metadata": {},
   "source": [
    "# Mice methylation data - Impute missing genes with mean"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "719df1c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "meth_df = pd.read_csv('../data/Cleveland/methylation_imputed.csv', index_col=0)\n",
    "\n",
    "mice_df = pd.read_csv('../data/Mice/data_methylation_hm450.csv',index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9a6a5405",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A2BP1</th>\n",
       "      <th>A2LD1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM7</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>psiTPTE22</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MEC1</th>\n",
       "      <td>0.78393</td>\n",
       "      <td>0.55490</td>\n",
       "      <td>0.74393</td>\n",
       "      <td>0.28846</td>\n",
       "      <td>0.19457</td>\n",
       "      <td>0.23815</td>\n",
       "      <td>0.58240</td>\n",
       "      <td>0.11336</td>\n",
       "      <td>0.13458</td>\n",
       "      <td>0.11345</td>\n",
       "      <td>...</td>\n",
       "      <td>0.09947</td>\n",
       "      <td>0.15068</td>\n",
       "      <td>0.13683</td>\n",
       "      <td>0.35715</td>\n",
       "      <td>0.70193</td>\n",
       "      <td>0.33222</td>\n",
       "      <td>0.26345</td>\n",
       "      <td>0.20008</td>\n",
       "      <td>0.18074</td>\n",
       "      <td>0.72662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M14</th>\n",
       "      <td>0.84968</td>\n",
       "      <td>0.27184</td>\n",
       "      <td>0.88745</td>\n",
       "      <td>0.55872</td>\n",
       "      <td>0.20010</td>\n",
       "      <td>0.27441</td>\n",
       "      <td>0.24414</td>\n",
       "      <td>0.14742</td>\n",
       "      <td>0.17265</td>\n",
       "      <td>0.13678</td>\n",
       "      <td>...</td>\n",
       "      <td>0.14274</td>\n",
       "      <td>0.19162</td>\n",
       "      <td>0.15129</td>\n",
       "      <td>0.32659</td>\n",
       "      <td>0.70403</td>\n",
       "      <td>0.32992</td>\n",
       "      <td>0.29607</td>\n",
       "      <td>0.21270</td>\n",
       "      <td>0.19070</td>\n",
       "      <td>0.40704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MDAMB134VI</th>\n",
       "      <td>0.84884</td>\n",
       "      <td>0.44047</td>\n",
       "      <td>0.58004</td>\n",
       "      <td>0.51544</td>\n",
       "      <td>0.16695</td>\n",
       "      <td>0.17913</td>\n",
       "      <td>0.43582</td>\n",
       "      <td>0.10513</td>\n",
       "      <td>0.12987</td>\n",
       "      <td>0.10102</td>\n",
       "      <td>...</td>\n",
       "      <td>0.09759</td>\n",
       "      <td>0.13494</td>\n",
       "      <td>0.09719</td>\n",
       "      <td>0.25859</td>\n",
       "      <td>0.63699</td>\n",
       "      <td>0.31373</td>\n",
       "      <td>0.27388</td>\n",
       "      <td>0.17380</td>\n",
       "      <td>0.14815</td>\n",
       "      <td>0.57372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MCC26</th>\n",
       "      <td>0.76965</td>\n",
       "      <td>0.39123</td>\n",
       "      <td>0.81830</td>\n",
       "      <td>0.40430</td>\n",
       "      <td>0.20149</td>\n",
       "      <td>0.23971</td>\n",
       "      <td>0.21317</td>\n",
       "      <td>0.12263</td>\n",
       "      <td>0.13767</td>\n",
       "      <td>0.11726</td>\n",
       "      <td>...</td>\n",
       "      <td>0.10332</td>\n",
       "      <td>0.16757</td>\n",
       "      <td>0.12046</td>\n",
       "      <td>0.33940</td>\n",
       "      <td>0.69271</td>\n",
       "      <td>0.34277</td>\n",
       "      <td>0.27065</td>\n",
       "      <td>0.18393</td>\n",
       "      <td>0.16858</td>\n",
       "      <td>0.50135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MCC13</th>\n",
       "      <td>0.77499</td>\n",
       "      <td>0.45356</td>\n",
       "      <td>0.79889</td>\n",
       "      <td>0.38255</td>\n",
       "      <td>0.20208</td>\n",
       "      <td>0.23711</td>\n",
       "      <td>0.23129</td>\n",
       "      <td>0.12546</td>\n",
       "      <td>0.13952</td>\n",
       "      <td>0.11780</td>\n",
       "      <td>...</td>\n",
       "      <td>0.11241</td>\n",
       "      <td>0.16395</td>\n",
       "      <td>0.12104</td>\n",
       "      <td>0.34009</td>\n",
       "      <td>0.67837</td>\n",
       "      <td>0.33435</td>\n",
       "      <td>0.27938</td>\n",
       "      <td>0.18776</td>\n",
       "      <td>0.16798</td>\n",
       "      <td>0.52900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICC3</th>\n",
       "      <td>0.79764</td>\n",
       "      <td>0.43894</td>\n",
       "      <td>0.81737</td>\n",
       "      <td>0.39870</td>\n",
       "      <td>0.17114</td>\n",
       "      <td>0.21492</td>\n",
       "      <td>0.19939</td>\n",
       "      <td>0.11488</td>\n",
       "      <td>0.13343</td>\n",
       "      <td>0.10639</td>\n",
       "      <td>...</td>\n",
       "      <td>0.09735</td>\n",
       "      <td>0.14092</td>\n",
       "      <td>0.10310</td>\n",
       "      <td>0.32754</td>\n",
       "      <td>0.67901</td>\n",
       "      <td>0.33334</td>\n",
       "      <td>0.24941</td>\n",
       "      <td>0.17248</td>\n",
       "      <td>0.16001</td>\n",
       "      <td>0.54184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICC9</th>\n",
       "      <td>0.68602</td>\n",
       "      <td>0.51602</td>\n",
       "      <td>0.79501</td>\n",
       "      <td>0.40044</td>\n",
       "      <td>0.16386</td>\n",
       "      <td>0.20731</td>\n",
       "      <td>0.26373</td>\n",
       "      <td>0.10301</td>\n",
       "      <td>0.13411</td>\n",
       "      <td>0.10000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.08782</td>\n",
       "      <td>0.12934</td>\n",
       "      <td>0.10061</td>\n",
       "      <td>0.26869</td>\n",
       "      <td>0.69777</td>\n",
       "      <td>0.30774</td>\n",
       "      <td>0.22858</td>\n",
       "      <td>0.17495</td>\n",
       "      <td>0.15519</td>\n",
       "      <td>0.64120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ICC12</th>\n",
       "      <td>0.64823</td>\n",
       "      <td>0.46904</td>\n",
       "      <td>0.79287</td>\n",
       "      <td>0.33122</td>\n",
       "      <td>0.19561</td>\n",
       "      <td>0.23294</td>\n",
       "      <td>0.23066</td>\n",
       "      <td>0.11852</td>\n",
       "      <td>0.14525</td>\n",
       "      <td>0.11275</td>\n",
       "      <td>...</td>\n",
       "      <td>0.10428</td>\n",
       "      <td>0.15722</td>\n",
       "      <td>0.11912</td>\n",
       "      <td>0.29987</td>\n",
       "      <td>0.72420</td>\n",
       "      <td>0.32530</td>\n",
       "      <td>0.25691</td>\n",
       "      <td>0.19706</td>\n",
       "      <td>0.17322</td>\n",
       "      <td>0.65651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCLF_PEDS_0003_T</th>\n",
       "      <td>0.82129</td>\n",
       "      <td>0.39638</td>\n",
       "      <td>0.81042</td>\n",
       "      <td>0.39977</td>\n",
       "      <td>0.19645</td>\n",
       "      <td>0.23987</td>\n",
       "      <td>0.28208</td>\n",
       "      <td>0.12768</td>\n",
       "      <td>0.13989</td>\n",
       "      <td>0.11926</td>\n",
       "      <td>...</td>\n",
       "      <td>0.10992</td>\n",
       "      <td>0.16441</td>\n",
       "      <td>0.11698</td>\n",
       "      <td>0.32357</td>\n",
       "      <td>0.66248</td>\n",
       "      <td>0.34323</td>\n",
       "      <td>0.28625</td>\n",
       "      <td>0.19521</td>\n",
       "      <td>0.18436</td>\n",
       "      <td>0.45919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ECC2</th>\n",
       "      <td>0.62260</td>\n",
       "      <td>0.45976</td>\n",
       "      <td>0.81694</td>\n",
       "      <td>0.36552</td>\n",
       "      <td>0.17408</td>\n",
       "      <td>0.21578</td>\n",
       "      <td>0.29433</td>\n",
       "      <td>0.11449</td>\n",
       "      <td>0.14623</td>\n",
       "      <td>0.10276</td>\n",
       "      <td>...</td>\n",
       "      <td>0.09997</td>\n",
       "      <td>0.13839</td>\n",
       "      <td>0.11243</td>\n",
       "      <td>0.28532</td>\n",
       "      <td>0.72184</td>\n",
       "      <td>0.30809</td>\n",
       "      <td>0.22911</td>\n",
       "      <td>0.18413</td>\n",
       "      <td>0.16672</td>\n",
       "      <td>0.70991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1521 rows × 14608 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     A1BG    A2BP1    A2LD1   A4GALT     AAAS     AACS  \\\n",
       "model_id                                                                 \n",
       "MEC1              0.78393  0.55490  0.74393  0.28846  0.19457  0.23815   \n",
       "M14               0.84968  0.27184  0.88745  0.55872  0.20010  0.27441   \n",
       "MDAMB134VI        0.84884  0.44047  0.58004  0.51544  0.16695  0.17913   \n",
       "MCC26             0.76965  0.39123  0.81830  0.40430  0.20149  0.23971   \n",
       "MCC13             0.77499  0.45356  0.79889  0.38255  0.20208  0.23711   \n",
       "...                   ...      ...      ...      ...      ...      ...   \n",
       "ICC3              0.79764  0.43894  0.81737  0.39870  0.17114  0.21492   \n",
       "ICC9              0.68602  0.51602  0.79501  0.40044  0.16386  0.20731   \n",
       "ICC12             0.64823  0.46904  0.79287  0.33122  0.19561  0.23294   \n",
       "CCLF_PEDS_0003_T  0.82129  0.39638  0.81042  0.39977  0.19645  0.23987   \n",
       "ECC2              0.62260  0.45976  0.81694  0.36552  0.17408  0.21578   \n",
       "\n",
       "                    AADAT    AAGAB     AAK1     AAMP  ...   ZSWIM7     ZW10  \\\n",
       "model_id                                              ...                     \n",
       "MEC1              0.58240  0.11336  0.13458  0.11345  ...  0.09947  0.15068   \n",
       "M14               0.24414  0.14742  0.17265  0.13678  ...  0.14274  0.19162   \n",
       "MDAMB134VI        0.43582  0.10513  0.12987  0.10102  ...  0.09759  0.13494   \n",
       "MCC26             0.21317  0.12263  0.13767  0.11726  ...  0.10332  0.16757   \n",
       "MCC13             0.23129  0.12546  0.13952  0.11780  ...  0.11241  0.16395   \n",
       "...                   ...      ...      ...      ...  ...      ...      ...   \n",
       "ICC3              0.19939  0.11488  0.13343  0.10639  ...  0.09735  0.14092   \n",
       "ICC9              0.26373  0.10301  0.13411  0.10000  ...  0.08782  0.12934   \n",
       "ICC12             0.23066  0.11852  0.14525  0.11275  ...  0.10428  0.15722   \n",
       "CCLF_PEDS_0003_T  0.28208  0.12768  0.13989  0.11926  ...  0.10992  0.16441   \n",
       "ECC2              0.29433  0.11449  0.14623  0.10276  ...  0.09997  0.13839   \n",
       "\n",
       "                    ZWINT     ZXDC   ZYG11A   ZYG11B      ZYX    ZZEF1  \\\n",
       "model_id                                                                 \n",
       "MEC1              0.13683  0.35715  0.70193  0.33222  0.26345  0.20008   \n",
       "M14               0.15129  0.32659  0.70403  0.32992  0.29607  0.21270   \n",
       "MDAMB134VI        0.09719  0.25859  0.63699  0.31373  0.27388  0.17380   \n",
       "MCC26             0.12046  0.33940  0.69271  0.34277  0.27065  0.18393   \n",
       "MCC13             0.12104  0.34009  0.67837  0.33435  0.27938  0.18776   \n",
       "...                   ...      ...      ...      ...      ...      ...   \n",
       "ICC3              0.10310  0.32754  0.67901  0.33334  0.24941  0.17248   \n",
       "ICC9              0.10061  0.26869  0.69777  0.30774  0.22858  0.17495   \n",
       "ICC12             0.11912  0.29987  0.72420  0.32530  0.25691  0.19706   \n",
       "CCLF_PEDS_0003_T  0.11698  0.32357  0.66248  0.34323  0.28625  0.19521   \n",
       "ECC2              0.11243  0.28532  0.72184  0.30809  0.22911  0.18413   \n",
       "\n",
       "                     ZZZ3  psiTPTE22  \n",
       "model_id                              \n",
       "MEC1              0.18074    0.72662  \n",
       "M14               0.19070    0.40704  \n",
       "MDAMB134VI        0.14815    0.57372  \n",
       "MCC26             0.16858    0.50135  \n",
       "MCC13             0.16798    0.52900  \n",
       "...                   ...        ...  \n",
       "ICC3              0.16001    0.54184  \n",
       "ICC9              0.15519    0.64120  \n",
       "ICC12             0.17322    0.65651  \n",
       "CCLF_PEDS_0003_T  0.18436    0.45919  \n",
       "ECC2              0.16672    0.70991  \n",
       "\n",
       "[1521 rows x 14608 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meth_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d070b209",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there are 249 missing genes\n",
      "there are 811 mis-cased genes\n"
     ]
    }
   ],
   "source": [
    "mice_genes = mice_df.index\n",
    "\n",
    "missing_genes = []\n",
    "for gene in meth_df.columns:\n",
    "    if (gene.upper() not in mice_genes):\n",
    "        missing_genes.append(gene)\n",
    "print(f\"there are {len(missing_genes)} missing genes\")\n",
    "\n",
    "miscased_genes = []\n",
    "for gene in meth_df.columns:\n",
    "    if (gene.upper() in mice_genes) and (gene not in mice_genes):\n",
    "        miscased_genes.append(gene)\n",
    "print(f\"there are {len(miscased_genes)} mis-cased genes\")\n",
    "miscased_genes[:5]\n",
    "\n",
    "# change mis-cased genes in the mice data:\n",
    "\n",
    "for gene in miscased_genes:\n",
    "    mice_df.rename(index={gene.upper():  gene}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "161331b2",
   "metadata": {},
   "source": [
    "fill the other missing genes with a mean value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f9ee9891",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mice dataset is (25865, 73)\n",
      "after insertions mice dataset is (26114, 73)\n"
     ]
    }
   ],
   "source": [
    "print(f\"mice dataset is {mice_df.shape}\")\n",
    "mice_samples = mice_df.columns\n",
    "\n",
    "rows_to_add = []\n",
    "for gene in missing_genes:\n",
    "    mean_value = meth_df[gene].mean()\n",
    "    row_values = pd.Series({sample: mean_value for sample in mice_samples}, name=gene)\n",
    "    rows_to_add.append(row_values)\n",
    "\n",
    "# Concatenate all new rows at once\n",
    "if rows_to_add:\n",
    "    new_rows_df = pd.DataFrame(rows_to_add)\n",
    "    mice_df = pd.concat([mice_df, new_rows_df], axis=0)\n",
    "\n",
    "print(f\"after insertions mice dataset is {mice_df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4089539f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mice_df.T.to_csv('data/mice/mice_methylation_imputed.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a704003",
   "metadata": {},
   "source": [
    "# Mice RNA - re-scale & impute missing genes with mean"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6d775d6",
   "metadata": {},
   "source": [
    "First: run rna_normalization.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "c7fdf885",
   "metadata": {},
   "outputs": [],
   "source": [
    "mice_df = pd.read_csv('data/mice_rna_normalized.csv', index_col=0)\n",
    "\n",
    "rna_df = pd.read_csv('../data/Cleveland/rna_imputed.csv', index_col=0)\n",
    "\n",
    "mosa_df = pd.read_csv('data/transcriptomics_used_by_MOSA.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "95bf2b52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "min: 1.17643791621576, max 15.2407734391336 mean 4.655333829072499 median 4.853046105867335\n",
      "negative values: 0\n",
      "(66, 14362)\n",
      "min: -11.39522, max 17.14427 mean 3.18137413668156 median 4.273594999999999\n",
      "negative values: 1567134\n",
      "(498, 15278)\n",
      "min: -12.39979, max 18.18559 mean 3.050093310831003 median 4.348235000000001\n",
      "negative values: 3559507\n",
      "(15320, 1072)\n"
     ]
    }
   ],
   "source": [
    "# Check if imputed RNA and (normalized) mice RNA are matching\n",
    "for df in [mice_df, rna_df, mosa_df]:\n",
    "    print(f\"min: {df.min().min()}, max {df.max().max()} mean {df.mean().mean()} median {df.median().median()}\")\n",
    "    print(f\"negative values: {df.lt(0).sum().sum()}\")\n",
    "    print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "fd840ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "values = list(rna_df.columns)\n",
    "\n",
    "with open(\"data/rna_imputed_list_of_genes.txt\", 'w') as output:\n",
    "    for row in values:\n",
    "        output.write(str(row) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "c812a354",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mice genes matching RNA genes: 14362\n",
      "Missing genes wrt RNA: 916\n"
     ]
    }
   ],
   "source": [
    "a = [mdf for mdf in mice_df.columns if mdf in rna_df.columns]\n",
    "print(f\"Mice genes matching RNA genes: {len(a)}\")\n",
    "\n",
    "a = [mdf for mdf in rna_df.columns if mdf not in mice_df.columns]\n",
    "print(f\"Missing genes wrt RNA: {len(a)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "177083b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "there are 916 missing genes\n",
      "there are 0 mis-cased genes\n"
     ]
    }
   ],
   "source": [
    "mice_genes = mice_df.columns\n",
    "\n",
    "missing_genes = []\n",
    "for gene in rna_df.columns:\n",
    "    if gene not in mice_genes:\n",
    "        missing_genes.append(gene)\n",
    "print(f\"there are {len(missing_genes)} missing genes\")\n",
    "\n",
    "miscased_genes = []\n",
    "for gene in rna_df.columns:\n",
    "    if (gene.upper() in mice_genes) and (gene not in mice_genes):\n",
    "        miscased_genes.append(gene)\n",
    "print(f\"there are {len(miscased_genes)} mis-cased genes\")\n",
    "miscased_genes[:5]\n",
    "\n",
    "# change mis-cased genes in the mice data:\n",
    "\n",
    "for gene in miscased_genes:\n",
    "    mice_df.rename(column={gene.upper():  gene}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "8403e4d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mice dataset is (66, 14362)\n",
      "after insertions mice dataset is (66, 15278)\n"
     ]
    }
   ],
   "source": [
    "print(f\"mice dataset is {mice_df.shape}\")\n",
    "\n",
    "means = {gene: float(rna_df[gene].mean()) for gene in missing_genes}\n",
    "\n",
    "new_cols_df = pd.DataFrame(index=mice_df.index, data=means)\n",
    "\n",
    "mice_df = pd.concat([mice_df, new_cols_df], axis=1)\n",
    "\n",
    "print(f\"after insertions mice dataset is {mice_df.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "7b924044",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mice genes matching RNA genes: 15278\n",
      "Missing genes wrt RNA: 0\n"
     ]
    }
   ],
   "source": [
    "a = [mdf for mdf in mice_df.columns if mdf in rna_df.columns]\n",
    "print(f\"Mice genes matching RNA genes: {len(a)}\")\n",
    "\n",
    "a = [mdf for mdf in rna_df.columns if mdf not in mice_df.columns]\n",
    "print(f\"Missing genes wrt RNA: {len(a)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "c50a886a",
   "metadata": {},
   "outputs": [],
   "source": [
    "mice_df.to_csv('data/mice/mice_rna_imputed.csv', index=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pnet-repro",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
